lib/parsers/ogg_parser.rb

```diff
@@ -0,0 +1,42 @@
+# https://xiph.org/vorbis/doc/Vorbis_I_spec.pdf
+# https://en.wikipedia.org/wiki/Ogg#Page_structure
+class FormatParser::OggParser
+  # The maximum size of an Ogg page is 65,307 bytes.
+  MAX_POSSIBLE_PAGE_SIZE = 65307
+
+  def call(io)
+    # The format consists of chunks of data each called an "Ogg page". Each page
+    # begins with the characters, "OggS", to identify the file as Ogg format.
+    capture_pattern = io.read(4)
+    return unless capture_pattern == 'OggS'
+
+    io.seek(28) # skip not important bytes
+
+    # Each header packet begins with the same header fields.
+    #   1) packet_type: 8 bit value (the identification header is type 1)
+    #   2) the characters v','o','r','b','i','s' as six octets
+    packet_type, vorbis, _vorbis_version, channels, sample_rate = io.read(16).unpack('Ca6LCL')
```

Comment:
`L` is "32-bit unsigned, native endian (uint32_t)". That means that if this code is run on a differently-endian platform it will give you results you do not expect. Granted, few people use PowerPC these days, and ARM has switchable endianness - but do you want to leave the user at the mercy of whatever their byte order is?

Reply:
I checked the docs and I found out that Ogg doesn't specify endianness:

> The Vorbis bitpacking convention specifies storage and bitstream manipulation at the byte, not word, level, thus host word ordering is of a concern only during optimization when writing high performance code that operates on a word of storage at a time rather than by byte. Logically, bytes are always coded and decoded in order from byte zero through byte n.
https://xiph.org/vorbis/doc/Vorbis_I_spec.pdf

 I am not 100% what it means and which endianness I should use.

Is there an easy way to test the parser on a differently-endian platform?

Reply:
In the section you are referring to the spec apparently describes the _bitstream_ of encoded audio, and for that section their statement makes sense as you do not need to specify endianness if you read byte-by-byte.

There is however a different section that _does_ specify endianness for encoded uints, let me know if you can't find it ;-) To make an experiment without having a platform of different endianness (or to be lazy and not having to read the spec) you can check which one is "native" for the platform you wrote the tests on in `irb` even:

```
'ohai'.unpack('L') == 'ohai'.unpack('V') # If true, your platform is V-native, if not - N-native
```